x-pack/plugins/observability/public/components/custom_threshold/components/preview_chart/preview_chart.tsx

```diff
@@ -0,0 +1,322 @@
+      }
+      let operationField = `${operation}(${sourceField})`;
+      if (metric?.filter) {
+        const aggFilter = JSON.stringify(metric.filter).replace(/"|\\/g, '');
+        operationField = `${operation}(${sourceField},kql='${aggFilter}')`;
+      }
+      return {
+        ...acc,
+        [metric.name]: operationField,
+      };
+    }, {} as AggMap);
+
+    setAggMap(aggMapFromMetrics);
+  }, [metrics]);
+
+  // Parse the equation
+  useEffect(() => {
+    try {
+      if (!aggMap) return;
+      const parser = new PainlessTinyMathParser({
+        aggMap,
+        equation: equation || Object.keys(aggMap || {}).join(' + '),
+      });
+
+      setFormula(parser.parse());
+    } catch (e) {
+      // The error will appear on Lens chart.
+      setAttributes(undefined);
+      return;
+    }
+  }, [aggMap, equation]);
+
+  const getOperationTypeFromRuleAggType = (aggType: CustomMetricAggTypes): OperationType => {
+    if (aggType === Aggregators.AVERAGE) return 'average';
+    if (aggType === Aggregators.CARDINALITY) return 'unique_count';
+    return aggType;
+  };
+
+  useEffect(() => {
+    if (!formulaAsync.value || !dataView || !formula) {
+      return;
+    }
+    const baseLayer = {
+      type: 'formula',
+      value: formula,
+      label: 'Custom Threshold',
+      groupBy,
+      format: {
+        id: 'number',
+        params: {
+          decimals: 0,
+        },
+      },
+      filter: {
+        language: 'kuery',
+        query: filterQuery || '',
+      },
+    };
+
+    const xYDataLayerOptions: XYLayerOptions = {
+      buckets: { type: 'date_histogram' },
+      seriesType: 'bar',
+    };
+
+    if (groupBy && groupBy?.length) {
+      xYDataLayerOptions.breakdown = {
+        type: 'top_values',
+        field: groupBy[0],
+        params: {
+          size: 3,
+          secondaryFields: groupBy as string[],
+          accuracyMode: false,
+        },
+      };
+    }
+
+    const xyDataLayer = new XYDataLayer({
+      data: [baseLayer].map((layer) => ({
+        type: layer.type,
+        value: layer.value,
+        label: layer.label,
+        format: layer.format,
+        filter: layer.filter,
+      })),
+      options: xYDataLayerOptions,
+    });
+
+    const layers: Array<XYDataLayer | XYReferenceLinesLayer> = [xyDataLayer];
+    if (thresholdReferenceLine) {
+      layers.push(...thresholdReferenceLine);
+    }
+    const attributesLens = new LensAttributesBuilder({
+      visualization: new XYChart({
+        visualOptions: {
+          valueLabels: 'hide',
+        },
+        layers,
+        formulaAPI: formulaAsync.value.formula,
+        dataView,
+      }),
+    }).build();
+    const lensBuilderAtt = { ...attributesLens, type: 'lens' };
+    setAttributes(lensBuilderAtt);
+  }, [
+    comparator,
+    dataView,
+    equation,
+    filterQuery,
+    formula,
+    formulaAsync.value,
+    groupBy,
+    metrics,
+    threshold,
+    thresholdReferenceLine,
+  ]);
+
+  if (
```

Comment:
What do you think about having a test for the error state?

Reply:
done